src/middlewared/middlewared/plugins/sysdataset.py

```diff
@@ -366,3 +370,151 @@ def __createlink(self, syspath, item):
                 await self.middleware.call('service.start', i)
 
         await self.__nfsv4link(config)
+
+    @private
+    def remove(self, path, link_only=False):
+        if os.path.exists(path):
+            if os.path.islink(path):
+                os.unlink(path)
+            elif os.path.isdir(path) and not link_only:
+                shutil.rmtree(path)
+            elif not link_only:
+                os.remove(path)
+
+    @private
+    def get_members(self, tar, prefix):
+        for tarinfo in tar.getmembers():
+            if tarinfo.name.startswith(prefix):
+                tarinfo.name = tarinfo.name[len(prefix):]
+                yield tarinfo
+
+    @private
+    def update_collectd_dataset(self):
+        config = self.middleware.call_sync('systemdataset.config')
+        is_freenas = self.middleware.call_sync('system.is_freenas')
+        rrd_mount = ''
+        if config['path']:
+            rrd_mount = f'{config["path"]}/rrd-{config["uuid"]}'
+
+        use_rrd_dataset = False
+        if (
+            rrd_mount and config['rrd'] and (
+                is_freenas or (not is_freenas and self.middleware.call_sync('notifier.failover_status') != 'BACKUP')
+            )
+        ):
+            use_rrd_dataset = True
+
+        # TODO: FreeNAS Config MD5 is hardcoded, this should be looked in future to see if we can change that
+        # If there is a failover table remove the rc.conf cache rc.conf.local will run again using the correct
+        # collectd_enable. See #5019
+        if is_freenas:
+            try:
+                os.remove('/var/tmp/freenas_config.md5')
+            except FileNotFoundError:
+                pass
+
+        hostname = self.middleware.call_sync('system.info')['hostname']
+        if not hostname:
+            hostname = self.middleware.call_sync('network.configuration.config')['hostname']
+
+        rrd_file = '/data/rrd_dir.tar.bz2'
+        data_dir = '/var/db/collectd/rrd'
+        disk_parts = psutil.disk_partitions()
+        data_dir_is_ramdisk = len([d for d in disk_parts if d.mountpoint == data_dir and d.device == 'tmpfs']) > 0
+
+        if use_rrd_dataset:
+            if os.path.isdir(rrd_mount):
+                if os.path.isdir(data_dir) and not os.path.islink(data_dir):
+                    if data_dir_is_ramdisk:
+                        # copy-umount-remove
+                        subprocess.Popen(
+                            f'cp -a "{data_dir}" "{data_dir}.{time.strftime("%Y%m%d%H%M%S")}"'.split(),
```

Comment:
We have talked about this before, split() is a bad example/precedent. Lets just use a list.